jumpscale/sals/nginx/nginx.py

```diff
@@ -113,27 +114,46 @@ def get_config(self):
         return render_config_template("website", base_dir=j.core.dirs.BASEDIR, website=self)
 
     def generate_certificates(self):
-        if self.ssl:
-            j.sals.process.execute(
-                f"certbot --nginx -d {self.domain} --non-interactive --agree-tos -m {self.letsencryptemail} --nginx-server-root {self.parent.cfg_dir}"
-            )
+        res = j.sals.process.execute(
+            f"certbot --nginx -d {self.domain} --non-interactive --agree-tos -m {self.letsencryptemail} --nginx-server-root {self.parent.cfg_dir}"
+        )
+        if res[0] != 0:
+            if self.selfsigned:
+                res = self.generate_self_signed_certificates()
+                if res and res[0] != 0:
+                    raise j.exceptions.JSException(f"Failed to generate self-signed certificate.{res}")
+                j.sals.fs.write_file(self.cfg_file, self.get_config())
+                return res
+            else:
+                raise j.exceptions.JSException(f"Failed to generate certificates by certbot.{res}")
+        return res
+
+    def generate_self_signed_certificates(self):
+        self.selfsigned = True
+        if j.sals.fs.exists(f"{self.parent.cfg_dir}/key.pem") and j.sals.fs.exists(f"{self.parent.cfg_dir}/cert.pem"):
+            return
+        res = j.sals.process.execute(
+            f"openssl req -nodes -x509 -newkey rsa:4096 -keyout {self.parent.cfg_dir}/key.pem -out {self.parent.cfg_dir}/cert.pem -days 365 -subj '/CN=localhost'"
+        )
+        return res
 
     def configure(self, generate_certificates=True):
         j.sals.fs.mkdir(self.cfg_dir)
-
         for location in self.get_locations():
             location.configure()
 
+        failback = self.selfsigned
+        self.selfsigned = False
         j.sals.fs.write_file(self.cfg_file, self.get_config())
+        self.selfsigned = failback
```

Comment:
i think you need to write the file after generating certs so you know if it's selfsigned or generated by certbot